From e73e4397ae6a67e14293c7502d74f16c5fb5f9c0 Mon Sep 17 00:00:00 2001
From: Oskar Fiedot <oskar.fiedot@intel.com>
Date: Tue, 27 Oct 2026 01:20:43 +0000
Subject: [PATCH 22/23] analyics modelinfo: Allow modelinfo file to have a
 single range

analyics modelinfo: Allow modelinfo file to have a single range

If the ModelInfo file has a single range, apply it to all channels when
doing the conversion to input and scales.

Part-of: <https://gitlab.freedesktop.org/gstreamer/gstreamer/-/merge_requests/11319>

(cherry picked from commit e09d21b70d98aba5945fee5fa46ec39047275ff1)
---
 .../gst-libs/gst/analytics/modelinfo.c        | 23 ++++++++++++-------
 1 file changed, 15 insertions(+), 8 deletions(-)

diff --git a/subprojects/gst-plugins-bad/gst-libs/gst/analytics/modelinfo.c b/subprojects/gst-plugins-bad/gst-libs/gst/analytics/modelinfo.c
index 908700af098..55a57ce7c0c 100644
--- a/subprojects/gst-plugins-bad/gst-libs/gst/analytics/modelinfo.c
+++ b/subprojects/gst-plugins-bad/gst-libs/gst/analytics/modelinfo.c
@@ -778,7 +778,8 @@ gst_analytics_modelinfo_get_input_scales_offsets (GstAnalyticsModelInfo *
 
   *output_scales = NULL;
   *output_offsets = NULL;
-  *num_output_ranges = 0;
+  if (num_output_ranges)
+    *num_output_ranges = 0;
 
   /* Get target ranges from modelinfo */
   if (!gst_analytics_modelinfo_get_target_ranges (modelinfo, tensor_name,
@@ -789,7 +790,7 @@ gst_analytics_modelinfo_get_input_scales_offsets (GstAnalyticsModelInfo *
   }
 
   /* Validate that input ranges match target ranges */
-  if (num_input_ranges != num_target_ranges) {
+  if (num_input_ranges != num_target_ranges && num_target_ranges != 1) {
     GST_ERROR
         ("Tensor '%s': number of input ranges (%zu) doesn't match number of "
         "target ranges in modelinfo (%zu)", tensor_name, num_input_ranges,
@@ -800,13 +801,18 @@ gst_analytics_modelinfo_get_input_scales_offsets (GstAnalyticsModelInfo *
   }
 
   /* Allocate output arrays */
-  *output_scales = g_new (gdouble, num_target_ranges);
-  *output_offsets = g_new (gdouble, num_target_ranges);
+  *output_scales = g_new (gdouble, num_input_ranges);
+  *output_offsets = g_new (gdouble, num_input_ranges);
 
   /* Calculate scale and offset for each channel */
-  for (i = 0; i < num_target_ranges; i++) {
-    target_min = target_mins[i];
-    target_max = target_maxs[i];
+  for (i = 0; i < num_input_ranges; i++) {
+    if (num_target_ranges == 1) {
+      target_min = target_mins[0];
+      target_max = target_maxs[0];
+    } else {
+      target_min = target_mins[i];
+      target_max = target_maxs[i];
+    }
     input_min = input_mins[i];
     input_max = input_maxs[i];
 
@@ -825,7 +831,8 @@ gst_analytics_modelinfo_get_input_scales_offsets (GstAnalyticsModelInfo *
         target_max, scale, offset);
   }
 
-  *num_output_ranges = num_target_ranges;
+  if (num_output_ranges)
+    *num_output_ranges = num_input_ranges;
   g_free (target_mins);
   g_free (target_maxs);
 
-- 
2.43.0

